### DISCUSSION TITLE
Confusion about `cacheGroup` and `priority`

### DISCUSSION THREAD

**Discussion by mxxk** [webpack/webpack#17056] [Q&A]
I seem to have encountered a strange (or at least, undocumented?) case with `SplitChunksPlugin`. Given this `webpack.config.js` snippet (see full example at https://replit.com/@mxxk/WebpackSplitChunksIdHints),

```js
module.exports = {
  ...
  optimization: {
    splitChunks: {
      chunks: "all",
      minSize: 0,
      cacheGroups: {
        // Disable default cache groups.
        default: false,
        defaultVendors: false,
        vendor: {
          priority: 5,
          enforce: true,
          reuseExistingChunk: true,
          test: nodeModulesRegexp,
        },
        main: {
          priority: 10,
          enforce: true,
          reuseExistingChunk: true,
          // `test` is omitted, so this cache group matches all modules.
        },
      },
    },
  },
};
```

I expected the cache group `main` to match all modules, and the cache group `vendor` to match only modules within `node_modules`. Because the `main` cache group has higher `priority`, I expected it to take precedence over `vendor`. (This example may seem pathological, but the original behavior was observed with the `defaultVendors` cache group before disabling it and adding an explicit `vendor` cache group.)

However, the actual result shows that the chunk containing `lodash` (a bundled dependency) is named [`main-vendor-node_modules_lodash_lodash_js.js`](https://180f2aa3-9b0a-4b00-b172-82326dc96a7f.id.repl.co/main-vendor-node_modules_lodash_lodash_js.js), which shows both `main` and `vendor` in its path. Note that `name` was intentionally omitted from both cache group entries to prevent Webpack from being forced to merge all matching modules into one chunk (for that cache group).

Output from compilation shows that the affected bundle has not one, but _two_ ID hints:

```
asset main-vendor-node_modules_lodash_lodash_js.js 532 KiB [emitted] (id hint: main, vendor)
```

The documentation for `priority`

> A module can belong to multiple cache groups. The optimization will prefer the cache group with a higher `priority`. The default groups have a negative priority to allow custom groups to take higher priority (default value is `0` for custom groups).

seems to suggest that cache groups with higher priorities take precedence over those with lower ones. However, in this case, it seems that _both_ cache groups were applied.

Is this correct / expected behavior? If so, can anyone offer an explanation as to why, so that this may be hopefully be better documented?

### Prior art

- https://github.com/webpack/webpack/discussions/15645

**Comment by alexander-akait** [comment]
You can give any names, it is just a name, i.e. `main-vendor-node_modules_lodash_lodash_js.js` was extracted from `main` entry and it is in `node_modules`

**Comment by mxxk** [comment]
Thank you @alexander-akait. It makes sense that the `idHint` (not in above config, but defaults to the chunk group key) can be explicitly set to an arbitrary string, and it's possible to change the name `main-vendor-node_modules_lodash_lodash_js.js` to something else. However, the fact that the name does contain both `main` and `vendor` tells me that _both_ `cacheGroup`s took effect on that chunk. I was hoping to get your help in understanding how that's possible, since my initial impression was that a _single_ `cacheGroup` is chosen for every module based on its `priority` relative to any other matching `cacheGroup`s.

I've read and debugged code in `SplitChunksPlugin.js`, but so far was unable to wrap my mind around the algorithm used there, so was wondering if you already have a good grasp on it and may be able to illuminate my confusion. 😄

So far, based on what I've observed, it seems more accurate to say that `priority` does not _prevent_ other `cacheGroup`s from matching, but rather it affects the order in which they are matched and processed (chunks potentially created), but my understanding is fairly limited beyond that.

**Reply by alexander-akait** [reply]
There is logic https://github.com/webpack/webpack/blob/main/lib/optimize/SplitChunksPlugin.js#L1316, it is on top as you can see, so we prefer chunk with `priority`, can you look which modules are not in place

**Reply by mxxk** [reply]
@alexander-akait I added some instrumentation to parts of `SplitChunksPlugin.js`, and am detailing what I found below. I'm not sure if this is helpful, but hope the findings lead us somewhere. Looks like if multiple cache groups match a module, they contribute their `idHint` values to the final chunk name, and this is intentional? 🤔 

---

First, it looks like this snippet

https://github.com/webpack/webpack/blob/e6f21b2ba259c4631c9f53c3b1093647c186377b/lib/optimize/SplitChunksPlugin.js#L1233-L1239

was being called three times:

| `cacheGroup.key` | `module.resource` |
| - | - |
| `main` | `webpack-split-chunks-id-hints/src/entry.js`
| `vendor` | `webpack-split-chunks-id-hints/node_modules/lodash/lodash.js`
| `main` | `webpack-split-chunks-id-hints/node_modules/lodash/lodash.js`

This shows `lodash` is being added twice, once per cache group. (Not sure what to make of this.)

---

At the start of the main loop,

https://github.com/webpack/webpack/blob/e6f21b2ba259c4631c9f53c3b1093647c186377b/lib/optimize/SplitChunksPlugin.js#L1307

`chunksInfoMap` contained two entries:

```js
"main chunks:da30707c": {
  "modules": [
    "webpack-split-chunks-id-hints/src/entry.js",
    "webpack-split-chunks-id-hints/node_modules/lodash/lodash.js"
  ],
  "cacheGroup": {
    "key": "main",
    "priority": 10,
    "minSize": {},
    "minChunks": 1,
    "idHint": "main",
    "reuseExistingChunk": true
  },
  "chunks": [
    {
      "idNameHints": [],
      "name": "main",
      "modules": [
        "webpack-split-chunks-id-hints/src/entry.js",
        "webpack-split-chunks-id-hints/node_modules/lodash/lodash.js"
      ]
    }
  ]
},
"vendor chunks:da30707c": {
  "modules": [
    "webpack-split-chunks-id-hints/node_modules/lodash/lodash.js"
  ],
  "cacheGroup": {
    "key": "vendor",
    "priority": 5,
    "minSize": {},
    "minChunks": 1,
    "idHint": "vendor",
    "reuseExistingChunk": true
  },
  "chunks": [
    {
      "idNameHints": [],
      "name": "main",
      "modules": [
        "webpack-split-chunks-id-hints/src/entry.js",
        "webpack-split-chunks-id-hints/node_modules/lodash/lodash.js"
      ]
    }
  ]
}
```

and indeed, based on `priority` the selected `bestEntry` corresponded to chunk group `main`. However, on the second iteration of the loop, there was only one entry corresponding to `vendor`

```js
"vendor chunks:da30707c": {
  "modules": [
    "webpack-split-chunks-id-hints/node_modules/lodash/lodash.js"
  ],
  "cacheGroup": {
    "key": "vendor",
    "priority": 5,
    "minSize": {},
    "minChunks": 1,
    "idHint": "vendor",
    "reuseExistingChunk": true
  },
  "chunks": [
    {
      "idNameHints": [
        "main"
      ],
      "name": "main",
      "modules": [
        "webpack-split-chunks-id-hints/src/entry.js",
        "webpack-split-chunks-id-hints/node_modules/lodash/lodash.js"
      ],
      "chunkReason": "reused as split chunk (cache group: main)"
    }
  ]
}
```

which became the `bestEntry`. This is how the string `"vendor"` was pushed to `idNameHints` resulting in `["main", "vendor"]`, and the final name observed in the original post.

**Reply by alexander-akait** [reply]
Yes, but what is the problem? :smile: Or it is just a question?

**Reply by mxxk** [reply]
Yes, I wasn't sure if there was anything wrong with the current behavior, which I was looking for help in understanding, so that's why I posted a discussion rather than an issue 😄
